Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Placement_wrappers.h

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2020  GeometryFactory (France). All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Fernando Cacciola <fernando.cacciola@geometryfactory.com>
+//                 Maxime Gimeno <maxime.gimeno@geometryfactory.com>
+
+#ifndef CGAL_SURFACE_MESH_SIMPLIFICATION_DETAIL_CONSTRAINED_PLACEMENT_H
+#define CGAL_SURFACE_MESH_SIMPLIFICATION_DETAIL_CONSTRAINED_PLACEMENT_H
+
+#include <CGAL/license/Surface_mesh_simplification.h>
+
+#include <CGAL/Surface_mesh_simplification/internal/Common.h>
+#include <CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Edge_profile.h>
+#include <CGAL/property_map.h>
+#include <CGAL/boost/graph/Named_function_parameters.h>
+
+#include <boost/optional.hpp>
+
+namespace CGAL {
+namespace Surface_mesh_simplification {
+namespace internal{
+
+template<class BasePlacement, class EdgeIsConstrainedMap>
+class Constrained_placement
+  : public BasePlacement
+{
+public:
+  typedef Tag_true Constrained_tag;
+  Constrained_placement(const EdgeIsConstrainedMap map,
+                        bool constrain_geometry,
+                        const BasePlacement& base = BasePlacement())
+    : BasePlacement(base),
+      m_ecm(map),
+      m_constrain_geom(constrain_geometry)
+  {}
+
+  template <typename Profile>
+  boost::optional<typename Profile::Point> operator()(const Profile& profile) const
+  {
+    if(!m_constrain_geom)
+    {
+      return BasePlacement::operator()(profile);
+    }
+    typedef typename Profile::TM                                    TM;
+    typedef typename boost::graph_traits<TM>::halfedge_descriptor   halfedge_descriptor;
+
+    for(halfedge_descriptor h : halfedges_around_target(profile.v0(), profile.surface_mesh()))
+    {
```

Comment:
@sloriot @afabri What should happen here if the base placement returns `boost::none` ? Right now, if one of the vertices is constrained, we just ignore the fact that the base failed and use that vertex, but should it just also return boost::none instead?

Reply:
In order not to surprise users, it should return `boost::none` as well IMO. Some users will use the cost, others the placement to prevent an edge to be collapsed.

Reply:
What about the cost ? It must have a cost to test every not constrained result when it was not done before.

Reply:
The cost is using the placement.

Reply:
yes